File: solutions/Arrays/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> st=new Stack<>();
        int maxA=0;
        int n=heights.length;
        for(int i=0;i<=n;i++)
        {
            while(!st.isEmpty() && (i==n || heights[st.peek()]>heights[i]))
            {   
                int height=heights[st.peek()];
                st.pop();
                int width;
                if(st.empty())
                {
                    width=i;
                }
                else
                {
                    width=i-st.peek()-1;
                }
                maxA=Math.max(maxA,height*width);
            }
            st.push(i);
        }
        return maxA;
    }
}

File: solutions/Arrays/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack=[]
        maxarea=0
        for i in range(len(heights)):
            while(stack and heights[stack[-1]]>heights[i]):
                element=stack.pop()
                nse=i
                pse= stack[-1] if stack else -1
                maxarea=max(maxarea,heights[element]*(nse-pse-1))

            stack.append(i)
        while stack:
                element=stack.pop()
                nse=len(heights)
                pse=stack[-1] if stack else -1
                maxarea=max(maxarea,heights[element]*(nse-pse-1))
        return maxarea
